src/awkward/operations/ak_type.py

```diff
@@ -64,15 +65,18 @@ def _impl(array):
     if array is None:
         return ak.types.UnknownType()
 
-    elif isinstance(
-        array,
-        tuple(x.type for x in ak.types.numpytype._dtype_to_primitive_dict),
+    elif isinstance(array, np.dtype):
+        return ak.types.NumpyType(ak.types.numpytype.dtype_to_primitive(array))
+
+    elif (
+        isinstance(array, np.generic)
+        or isinstance(array, builtins.type)
+        and issubclass(array, np.generic)
     ):
-        return ak.types.NumpyType(
-            ak.types.numpytype._dtype_to_primitive_dict[array.dtype]
-        )
+        primitive = ak.types.numpytype.dtype_to_primitive(np.dtype(array))
+        return ak.types.NumpyType(primitive)
 
-    elif isinstance(array, (bool, np.bool_)):
+    elif isinstance(array, bool):
```

Comment:
I have to revert this:

```suggestion
    elif isinstance(array, (bool, np.bool_)):
```

because a NumPy boolean is not a Python boolean:

```python
>>> [isinstance(x, bool) for x in np.array([False, True])]
[False, False]
>>> [isinstance(x, np.bool_) for x in np.array([False, True])]
[True, True]
```

Reply:
The `isinstance(x, np.generic)` should pick up the instances of the primitive NumPy `type`s:
```pycon
>>> isinstance(np.array([True,False])[0], np.generic)
True
```

Reply:
That's true. And I see that the `np.generic` check is above the boolean one. Well, checking for `(bool, np.bool_)` is common enough to be an idiom, and it's paired with checking numbers for `numbers.Integral` and `numbers.Real` (usually written that way to catch NumPy, though it would catch more than that).